heudiconv/bids.py

```diff
@@ -0,0 +1,323 @@
+        "study, possibly pointing to pre-registration (if public or embargoed)")
+
+    # TODO: collect all task- .json files for func files to
+    tasks = {}
+    # way too many -- let's just collect all which are the same!
+    # FIELDS_TO_TRACK = {'RepetitionTime', 'FlipAngle', 'EchoTime',
+    #                    'Manufacturer', 'SliceTiming', ''}
+    for fpath in find_files('.*_task-.*\_bold\.json', topdir=path,
+                        exclude_vcs=True, exclude="/\.(datalad|heudiconv)/"):
+        task = re.sub('.*_(task-[^_\.]*(_acq-[^_\.]*)?)_.*', r'\1', fpath)
+        json_ = load_json(fpath)
+        if task not in tasks:
+            tasks[task] = json_
+        else:
+            rec = tasks[task]
+            # let's retain only those fields which have the same value
+            for field in sorted(rec):
+                if field not in json_ or json_[field] != rec[field]:
+                    del rec[field]
+        # create a stub onsets file for each one of those
+        suf = '_bold.json'
+        assert fpath.endswith(suf)
+        events_file = fpath[:-len(suf)] + '_events.tsv'
+        lgr.debug("Generating %s", events_file)
+        with open(events_file, 'w') as f:
+            f.write("onset\tduration\ttrial_type\tresponse_time")
+    # extract tasks files stubs
+    for task_acq, fields in tasks.items():
+        task_file = op.join(path, task_acq + '_bold.json')
+        lgr.debug("Generating %s", task_file)
+        fields["TaskName"] = ("TODO: full task name for %s" %
+                              task_acq.split('_')[0].split('-')[1])
+        fields["CogAtlasID"] = "TODO"
+        with open(task_file, 'w') as f:
+            f.write(json_dumps_pretty(fields, indent=2, sort_keys=True))
+
+
+def tuneup_bids_json_files(json_files):
+    """Given a list of BIDS .json files, e.g. """
+    if not json_files:
+        return
+    # Harmonize generic .json formatting
+    for jsonfile in json_files:
+        json_ = load_json(jsonfile)
+        # sanitize!
+        for f1 in ['Acquisition', 'Study', 'Series']:
+            for f2 in ['DateTime', 'Date']:
+                json_.pop(f1 + f2, None)
+        # TODO:  should actually be placed into series file which must
+        #        go under annex (not under git) and marked as sensitive
+        # MG - Might want to replace with flag for data sensitivity
+        # related - https://github.com/nipy/heudiconv/issues/92
+        if 'Date' in str(json_):
+            # Let's hope no word 'Date' comes within a study name or smth like
+            # that
+            raise ValueError("There must be no dates in .json sidecar")
+        #json.dump(json_, open(jsonfile, 'w'), indent=2)
+        save_json(jsonfile, json_, indent=2) # ensure this does same as above
+
+    # Load the beast
+    seqtype = op.basename(op.dirname(jsonfile))
+
+    # MG - want to expand this for other _epi
+    # possibly add IntendedFor automatically as well?
```

Comment:
thought about it, but not sure if we could easily pull it off since it is not 100% clear what file(s) fmap is pertinent to.  We could do matching by size/encoding directions etc, but it would be truly a heuristic ;)  may be should become some kind of an option ("--indend-fmaps-for-all-epis"? ;))

Reply:
technically the shim and encoding direction (at least one) has to remain the same. i think some of this could be done automatically. perhaps we can generate this as a function that people can reuse in the heuristic. in fact it may be useful, now that this is a proper python package to create a set of utility functions that simplify heuristics. also would be useful to ensure that json files can be changed via heuristics as well.